tools/load_system_tablets.cc

```diff
@@ -0,0 +1,244 @@
+/*
+ * Copyright (C) 2024-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#include "tools/load_system_tablets.hh"
+
+#include <algorithm>
+#include <seastar/core/thread.hh>
+#include <seastar/util/closeable.hh>
+
+#include "log.hh"
+#include "partition_slice_builder.hh"
+#include "data_dictionary/keyspace_metadata.hh"
+#include "db/cql_type_parser.hh"
+#include "db/system_keyspace.hh"
+#include "mutation/mutation.hh"
+#include "readers/combined.hh"
+#include "reader_concurrency_semaphore.hh"
+#include "tools/sstable_manager_service.hh"
+#include "types/list.hh"
+#include "types/tuple.hh"
+
+namespace {
+
+logging::logger logger{"load_sys_tablets"};
+
+
+future<std::filesystem::path> get_table_directory(std::filesystem::path scylla_data_path,
+                                                  std::string_view keyspace_name,
+                                                  std::string_view table_name) {
+    logger.trace("get_table_directory: {}.{} in {}", keyspace_name, table_name, scylla_data_path);
+    auto system_tables_path = scylla_data_path / keyspace_name;
+    auto system_tables_dir = co_await open_directory(system_tables_path.native());
+    sstring found;
+    // locate a directory named like
+    //   "$scylla_data_path/system/tablets-fd4f7a4696bd3e7391bf99eb77e82a5c"
+    auto h = system_tables_dir.list_directory([&] (directory_entry de) -> future<> {
+        if (!found.empty()) {
+            return make_ready_future();
+        }
+
+        auto dash_pos = de.name.find_last_of('-');
+        if (dash_pos == de.name.npos) {
+            // unlikely. but this should not be fatal
+            return make_ready_future();
+        }
+        if (de.name.substr(0, dash_pos) != table_name) {
+            return make_ready_future();
+        }
+        if (!de.type) {
+            throw std::runtime_error(fmt::format("failed to load system.tablets from {}/{}: unrecognized type", scylla_data_path, de.name));
+        }
+        if (*de.type != directory_entry_type::directory) {
+            throw std::runtime_error(fmt::format("failed to load system.tablets from {}/{}: not a directory", scylla_data_path, de.name));
+        }
+        found = de.name;
+        return make_ready_future();
+    });
+    co_await h.done();
+
+    if (found.empty()) {
+        throw std::runtime_error(fmt::format("failed to load system.tablets from {}: couldn't find table directory", scylla_data_path));
+    }
+    co_return system_tables_path / found.c_str();
+}
+
+mutation_opt read_mutation(sharded<sstable_manager_service>& sst_man,
```

Comment:
This is a lot of duplication with the schema loader code. We could generalize this to a single method:
`mutation_opt read_mutation_from_table_offline()`
This would get a partition key and (optionally) a clustering key and return the mutation. Then the different callers would just need to prepare the keys and process the resulting mutation.

Reply:
ack. extracted.